wsmux/session.go

```diff
@@ -0,0 +1,334 @@
+		case <-s.closed:
+			return
+
+		case fr, ok := <-s.writes:
+			if !ok {
+				s.logger.Print("write channel closed")
+				return
+			}
+			err := s.conn.WriteMessage(websocket.BinaryMessage, fr.Write())
+			s.logger.Printf("wrote message: %v", fr)
+			if err != nil {
+				s.logger.Print(err)
+			}
+		}
+	}
+}
+
+func (s *Session) recvLoop() {
+	for {
+		select {
+		case <-s.closed:
+			return
+		default:
+		}
+
+		t, msgReader, err := s.conn.NextReader()
+		if err != nil {
+			_ = s.Close()
+			break
+		}
+		if t != websocket.BinaryMessage {
+			continue
+		}
+
+		h, err := getHeader(msgReader)
+		id, msgType := h.id(), h.msg()
+
+		switch msgType {
+		//Used for creating a new stream
+		case msgSYN:
+			s.mu.Lock()
+			if _, ok := s.streams[id]; ok {
+				s.logger.Printf("received duplicate syn frame for stream %d", id)
+				s.mu.Unlock()
+				break
+			}
+
+			str := newStream(id, s)
+			// no point in locking here
+			close(str.accepted)
+
+			s.streams[id] = str
+			select {
+			case s.writes <- newAckFrame(id, uint32(DefaultCapacity)):
+				s.logger.Printf("sent ack frame")
+			default:
+				s.logger.Printf("Error accepting stream")
+				s.mu.Unlock()
+				return
+			}
+			s.asyncPushStream(str)
+			s.mu.Unlock()
+
+		//received data
+		case msgDAT:
+			s.mu.Lock()
+			str, ok := s.streams[id]
+			s.mu.Unlock()
+			if !ok {
+				s.logger.Printf("received data frame for unknown stream %d", id)
+				break
+			}
+			b, err := ioutil.ReadAll(msgReader)
+			if err != nil {
+				s.logger.Print(err)
+				break
+			}
+			str.addToBuffer(b)
+			str.notifyRead()
+			s.logger.Printf("received DAT frame on stream %d: %v", id, bytes.NewBuffer(b))
+
+		//received ack frame
+		case msgACK:
+			s.mu.Lock()
+			str, ok := s.streams[id]
+			s.mu.Unlock()
+			s.logger.Printf("received ack frame")
+			if !ok {
+				s.logger.Printf("received ack frame for unknown stream %d", id)
+				break
+			}
+
+			b := make([]byte, 4)
+			_, err := msgReader.Read(b)
+			if err != nil {
+				s.logger.Print(err)
+				break
+			}
+			read := binary.LittleEndian.Uint32(b)
+			select {
+			case <-str.accepted:
+			default:
+				// close str.accepted to accept stream
+				s.logger.Printf("accepting stream")
+				close(str.accepted)
+			}
+			s.logger.Printf("received ack frame: id %d: remote read %d bytes", id, read)
+			str.updateRemoteCapacity(read)
+			str.notifyWrite()
+
+		// received fin frame
+		case msgFIN:
+			s.mu.Lock()
+			str, ok := s.streams[id]
+			s.mu.Unlock()
+			if !ok {
+				s.logger.Printf("received fin frame for unknown stream %d", id)
+				break
+			}
+
+			err := str.setRemoteClosed()
+			if str != nil {
+				s.logger.Print(err)
+			}
+			s.logger.Printf("remote stream %d closed connection", id)
+		}
+
+	}
+}
+
+func (s *Session) removeStream(id uint32) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if _, ok := s.streams[id]; ok {
+		delete(s.streams, id)
+	}
+}
+
+func (s *Session) asyncPushStream(str *stream) {
+	select {
+	case s.streamCh <- str:
+	default:
```

Comment:
Instead of a queue length how about a timeout?
The do `go asyncPushStream(str)`

Hmmm, maybe we should think the pro/cons...